### DISCUSSION TITLE
use spng for decoding PNGs and libpng only for encoding

### DISCUSSION THREAD

**Discussion by dloebl** [libvips/libvips#2900] [General]
We noticed that `spng` seems to be faster at decoding PNGs but slower than `libpng` at encoding.
The timings below mean the total time it took to load,process and save an image with libvips:
| **nsamples** | **source image type** | **target image type** | **`spng v0.7.2` avg in _ms_** | **`libpng v1.6.37` avg in _ms_** | **`spng v0.7.2` p99 in _ms_** | **`libpng v1.6.37` p99 in _ms_** |
| - | - | - | - | - | - | - |
| ~29000 | image/png | image/png | 472 | 431 | 3374 | 2933 | 
| ~44000 | image/png | `!=` image/png | 375 | 389 | 2743 | 2748 |
| ~180 | `!=` image/png | image/png | 424 | 346 | 1559 | 1488 |

**zlib version:** zlib-ng v2.0.6
**libvips version:** a129476f0febbca8e0df9c6766b141047a20e89c

Is it possible to link against spng/libpng at the same time and use spng just for decoding and libpng only for encoding PNGs?

**Comment by jcupitt** [comment]
Hi @dloebl,

That's interesting. 

I think 90%+ of CPU time in PNG save is spent in libz, so I would guess you are somehow using two different libzes in your tests. Confusingly:

```
$ ldd $(which vips) | grep libz
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fb35d771000)
	libznz.so.3 => /lib/x86_64-linux-gnu/libznz.so.3 (0x00007fb35be69000)
	libzstd.so.1 => /lib/x86_64-linux-gnu/libzstd.so.1 (0x00007fb35b4c6000)
```

What a mess. Have you tried in callgrind?

**Reply by jcupitt** [reply]
I tried quickly in callgrind here.

First, using libpng and a 1450 x 2048 pixel test image, libvips makes four calls to `png_write_rows()` and that burns 95.7% of CPU. 

libpng in turn makes 2048 calls off into libz.so.1.2.11, and that uses 78% of CPU time, so 17% in libpng.

Second, using git master libspng, libvips makes 2048 calls to spng_encode_row() using 96.4% of CPU time. 

That in turn makes 2262 calls into libz.so.1.2.11, using 65% of CPU. libspng's `encode_scanline()` is using 31.4% of CPU.

So yes, it does look like libpng is lower overhead than libspng for this case.

@randy408, what do you think?

**Reply by randy408** [reply]
Are the file sizes identical? Quick way to check if settings are the same

**Reply by jcupitt** [reply]
Yes, to the byte.

**Reply by randy408** [reply]
Although the different number of zlib calls is weird it's probably filtering that's slower, if I had to guess.

**Reply by jcupitt** [reply]
Good point, I'd forgotten about filtering.

With `filter=none`, libpng and libspng seem to be about the same speed. The libvips default is `filter=all`, and on that libspng seems to be maybe 10% slower on my test image (a big photo).

**Reply by kleisauke** [reply]
It seems that libspng spends more CPU time finding the best row filter when `filter=all` is used, see e.g. these 2 flamegraphs:
https://github.com/kleisauke/vips-flamegraph/raw/master/output/bench-spng.svg - libspng
https://github.com/kleisauke/vips-flamegraph/raw/master/output/bench-libpng.svg - libpng

Instructions for producing these flamegraphs can be found in the https://github.com/kleisauke/vips-flamegraph repository.

**Reply by randy408** [reply]
Is it more time in `get_best_filter` compared to libpng's `png_write_find_filter` or filtering overall (finding the best filter + filtering itself)? I don't see any time data.

**Reply by kleisauke** [reply]
IIUC, in that flamegraph, libpng spends ~1.3% of CPU on finding the best filter (CPU time of `png_write_find_filter` minus the above call stacks), ~13.3% on the filtering (the `png_setup_<FILTER>_row` functions), and ~79.0% on writing the filtered rows (the `png_write_filtered_row` function).

libspng on the other hand, spends ~20.2% on finding the best filter (the `get_best_filter` function), ~7.2% on the filtering (the `filter_scanline` function), and ~67.0% on writing the filtered rows (the `write_idat_bytes` function).

So for the overall filtering (finding the best filter + filtering itself), libspng spends ~27.4% CPU, while libpng spends ~14.6% CPU. In terms of timing, the ~13% difference can also be noticed when the adaptive filter is avoided, see e.g:
```bash
# libpng filter=all
$ /usr/bin/time -f %M:%e vips copy images/x.jpg output/x2.png
43192:3.97
# libpng filter=paeth
$ /usr/bin/time -f %M:%e vips copy images/x.jpg output/x2.png[filter=paeth]
43200:3.54

# libspng filter=all
$ /usr/bin/time -f %M:%e vips copy images/x.jpg output/x2.png
43416:4.59
# libspng filter=paeth
$ /usr/bin/time -f %M:%e vips copy images/x.jpg output/x2.png[filter=paeth]
42668:3.53
```

(Note that those flamegraphs are best viewed locally, as they are not interactive when served via GitHub)

**Reply by randy408** [reply]
Eventually filtering will be optimized using SIMD intrinsics for x86 and ARM. Right now both libraries use plain C code to achieve the same thing, compilers appear to handle libpng's code better.
